2023/YoctoSummit-2023.11/fdo/fido-device-onboarding.md

```diff
@@ -0,0 +1,516 @@
+- Turns out cross-compiling rust is harder than it seemed at first
+
+---
+
+# State of rust in Yocto
+
+<br>
+<br>
+
+.left-column50[
+
+### meta-rust
+
+- Source-based, thus more customizable
+- Allows for an offline build via `cargo-bitbake`
+- Doesn't contain `cross` or `rustc-nightly`
+]
+
+.right-column50[
+
+### meta-rust-bin
+
+- Provides pre-built compiler and `cargo`
+- Needs online build
+- Allows for the use of `rustc-nightly` and `cross`
+]
+
+???
+
+- Both exist to satisfy different requirements
+- `cargo-bitbake` generates .bb recipe from `Cargo.toml`
+    + Yocto also has the `cargo-update-recipe-crates`
+        - It allows for updating the cargo dependencies via bitbake -c
+        update_crates recipe
+            + Wasn't used by us as the project was done for the Honister version
+- nightly builds of rustc contain some improvements to cross-compilation that
+will be upstreamed in the future
+    + Besides some improvements to parallelism and multi-threading it adds
+    `Cross` crate
+        - Contenerized toolchain and linker
+        - Theoretically should work
+        - Abandoned this trope as it seemed less stable
+
+---
+
+# Bitbake environment vs pkg-config
+
+<br>
+<br>
+
+## .center[pkg-config]
+
+_The `pkg-config` command usually doesn’t support cross-compilation, and this
+crate prevents it from selecting incompatible versions of libraries. Setting
+`PKG_CONFIG_ALLOW_CROSS=1` disables this protection,
+**which is likely to cause linking errors**, unless `pkg-config` has been
+configured to use appropriate sysroot and search paths for the target platform._
+
+[docs.rs/pkg_config](https://docs.rs/pkg-config/latest/pkg_config/)
+
+???
+
+- The Host dependencies were also built using the pkg-config and used aarch64
+flags
+- The reverse happened for linker
+[meta-rust issue#431](https://github.com/meta-rust/meta-rust/issues/431)
+
+---
+
+# Bitbake environment vs pkg-config
+
+<br>
+<br>
+
+```bitbake
+# By default pkg-config variables point to aarch64 libraries which are picked up
+# during x86_64 builds, this causes aarch64 include directories and linker
+# search paths to into x86_64 builds, causing problems.
+#
+# Host libraries already use absolute paths so set sysroot to /
+export PKG_CONFIG_SYSROOT_DIR="/"
+export PKG_CONFIG_PATH="${RECIPE_SYSROOT_NATIVE}/usr/lib/pkgconfig:${RECIPE_SYSROOT_NATIVE}/usr/share/pkgconfig"
+export PKG_CONFIG_LIBDIR="${RECIPE_SYSROOT_NATIVE}/usr/lib/pkgconfig"
+export PKG_CONFIG_DIR="${RECIPE_SYSROOT_NATIVE}/usr/lib/pkgconfig"
+
+# Those variables are handled internally by pkg-config crate.
+# All paths are relative to sysroot, so set PKG_CONFIG_SYSROOT_DIR
+export PKG_CONFIG_SYSROOT_DIR_${TARGET_SYS}="${RECIPE_SYSROOT}"
+export PKG_CONFIG_PATH_${TARGET_SYS}="${RECIPE_SYSROOT}/usr/lib/pkgconfig:${RECIPE_SYSROOT}/usr/share/pkgconfig"
+export PKG_CONFIG_LIBDIR_${TARGET_SYS}="${RECIPE_SYSROOT}/usr/lib/pkgconfig"
+export PKG_CONFIG_DIR_${TARGET_SYS}="${RECIPE_SYSROOT}/usr/lib/pkgconfig"
```

Comment:
Does not fit the presentation width maybe try `.code-small` or sth like that, check slides.css classes.

Reply:
Renders now properly with `.code-13px` class [link](https://github.com/3mdeb/conferences/blob/yocto_summit_2023_fdo/2023/YoctoSummit-2023.11/fdo/fido-device-onboarding.md?plain=1#L374)